### PR TITLE
[2.7.0] bounded int libfuncs

### PR DESCRIPTION
**Pull request by edg-l**
## Checklist
- [ ] Linked to Github Issue
- [ ] Unit tests added
- [ ] Integration tests added.
- [ ] This change requires new documentation.
  - [ ] Documentation has been added/updated.

### CODE REVIEW

**Comment by edg-l** [comment]
Status:

```
failures:
   core::test::integer_test::test_i8_operators - Panicked with "assertion failed: `19_i8 / -7_i8 == -2_i8`.".
   core::test::integer_test::test_i8_divmod_overflow - expected panic but finished successfully.
   core::test::integer_test::test_i16_operators - Panicked with "assertion failed: `19_i16 / -7_i16 == -2_i16`.".
   core::test::integer_test::test_i16_divmod_overflow - expected panic but finished successfully.
   core::test::integer_test::test_i32_operators - Panicked with "assertion failed: `19_i32 / -7_i32 == -2_i32`.".
   core::test::integer_test::test_i32_divmod_overflow - expected panic but finished successfully.
   core::test::integer_test::test_i64_operators - Panicked with "assertion failed: `19_i64 / -7_i64 == -2_i64`.".
   core::test::integer_test::test_i64_divmod_overflow - expected panic but finished successfully.
   core::test::integer_test::test_i128_operators - Panicked with "assertion failed: `19_i128 / -7_i128 == -2_i128`.".
   core::test::integer_test::test_i128_divmod_overflow - expected panic but finished successfully.
   core::test::integer_test::bounded_int::test_add - Panicked with "assertion failed: `upcast(bounded_int_add(-128_i8, -128_i8)) == -256_felt252`.".
   core::test::integer_test::bounded_int::test_sub - Panicked with "assertion failed: `upcast(bounded_int_sub(0_u8, 255_u8)) == -255_felt252`.".
   core::test::integer_test::bounded_int::test_mul - Panicked with "assertion failed: `upcast(bounded_int_mul(-128_i8, 127_i8)) == -128_felt252 * 127`.".
   core::test::integer_test::bounded_int::test_constrain - Panicked with "assertion failed: `test_constrain_helper::<u8, 0x80>(0x80)`.".

Error: test result: FAILED. 438 passed; 14 failed; 3 ignored
make: *** [test-cairo] Error 1
```

File: src/libfuncs/bounded_int.rs [hunk 1]
```diff
@@ -0,0 +1,739 @@
...
+            .clone(),
+        upper: (&lhs_range.upper)
+            .max(&rhs_range.upper)
+            .max(&dst_range.upper)
+            .clone(),
+    };
+    let compute_ty = IntegerType::new(context, compute_range.offset_bit_width()).into();
+
+    // Zero-extend operands into the computation range.
+    assert!(compute_range.offset_bit_width() >= lhs_width);
```
Comment (pefontana) [inline src/libfuncs/bounded_int.rs:116]: Can we return an error here?

File: src/libfuncs/bounded_int.rs [hunk 2]
```diff
@@ -0,0 +1,739 @@
...
+    // Extract the ranges for the operands and the result type.
+    let lhs_ty = registry.get_type(&info.signature.param_signatures[0].ty)?;
+    let rhs_ty = registry.get_type(&info.signature.param_signatures[1].ty)?;
+
+    let lhs_range = lhs_ty.integer_range(registry).unwrap();
+    let rhs_range = rhs_ty.integer_range(registry).unwrap();
+    let dst_range = registry
+        .get_type(&info.signature.branch_signatures[0].vars[0].ty)?
+        .integer_range(registry)
+        .unwrap();
```
Comment (pefontana) [inline src/libfuncs/bounded_int.rs:211]: Remove unwrap

File: src/libfuncs/bounded_int.rs [hunk 3]
```diff
@@ -0,0 +1,739 @@
...
+            .clone(),
+        upper: (&lhs_range.upper)
+            .max(&rhs_range.upper)
+            .max(&dst_range.upper)
+            .clone(),
+    };
+    let compute_ty = IntegerType::new(context, compute_range.offset_bit_width()).into();
+
+    // Zero-extend operands into the computation range.
+    assert!(compute_range.offset_bit_width() >= lhs_width);
```
Comment (pefontana) [inline src/libfuncs/bounded_int.rs:238]: same here,
Can we return and error and avoid the assert! ?

File: src/libfuncs/bounded_int.rs [hunk 4]
```diff
@@ -0,0 +1,739 @@
...
+    let src_width = if src_ty.is_bounded_int(registry) {
+        src_range.offset_bit_width()
+    } else {
+        src_range.zero_based_bit_width()
+    };
+
+    let lower_range = registry
+        .get_type(&info.branch_signatures()[0].vars[1].ty)?
+        .integer_range(registry)
+        .unwrap();
```
Comment (pefontana) [inline src/libfuncs/bounded_int.rs:591]: remove unwrap()

File: src/libfuncs/bounded_int.rs [hunk 5]
```diff
@@ -0,0 +1,739 @@
...
+    let upper_range = registry
+        .get_type(&info.branch_signatures()[1].vars[1].ty)?
+        .integer_range(registry)
+        .unwrap();
```
Comment (pefontana) [inline src/libfuncs/bounded_int.rs:595]: remove unwrap()

File: src/libfuncs/bounded_int.rs [hunk 6]
```diff
@@ -0,0 +1,739 @@
...
+    helper: &LibfuncHelper<'ctx, 'this>,
+    _metadata: &mut MetadataStorage,
+    info: &SignatureOnlyConcreteLibfunc,
+) -> Result<()> {
+    let src_value: Value = entry.argument(0)?.into();
+
+    let src_ty = registry.get_type(&info.signature.param_signatures[0].ty)?;
+    let src_range = src_ty.integer_range(registry).unwrap();
+
+    assert!(
```
Comment (pefontana) [inline src/libfuncs/bounded_int.rs:695]: return error

**Review by pefontana — CHANGES_REQUESTED** [review]
Nice one @edg-l @azteca1998 !!!!
@azteca1998, can you remove the unwraps and asserts and replace them with Errors returns

**Review by pefontana — CHANGES_REQUESTED** [review]
@azteca1998 Can you fix the coverage CI, I think it is failing because of the circuits corelib tests
Also, can we add some unit tests for the new libfuncs?

**Review by pefontana — APPROVED** [review]

File: src/libfuncs/cast.rs [hunk 7]
```diff
@@ -57,200 +55,205 @@ pub fn build_downcast<'ctx, 'this>(
...
+    if info.signature.param_signatures[1].ty == info.signature.branch_signatures[0].vars[1].ty {
+        let k0 = entry.const_int(context, location, 0, 1)?;
+        entry.append_operation(helper.cond_br(
+            context,
+            k0,
+            [0, 1],
+            [&[range_check, src_value], &[range_check]],
+            location,
+        ));
+        return Ok(());
```
Comment (JulianGCalderon) [inline src/libfuncs/cast.rs:74]: Out of curiosity, why is `cond_br` necessary when the condition is constant?

Comment (azteca1998) [reply src/libfuncs/cast.rs:74]: The compiler expects to have a `cf.br` operation for each branch in the libfunc. If we didn't have this, cairo native would throw an error at runtime. LLVM should optimize it away anyways.

Comment (JulianGCalderon) [reply src/libfuncs/cast.rs:74]: Aah, thanks

File: src/libfuncs/bounded_int.rs [hunk 8]
```diff
@@ -0,0 +1,739 @@
...
+        ))?
+    } else {
+        res_value
+    };
+
+    entry.append_operation(helper.br(0, &[res_value], location));
+    Ok(())
+}
+
+/// Generate MLIR operations for the `bounded_int_add` libfunc.
```
Comment (JulianGCalderon) [inline src/libfuncs/bounded_int.rs:188]:
```suggestion
/// Generate MLIR operations for the `bounded_int_sub` libfunc.
```

File: src/libfuncs/bounded_int.rs [hunk 9]
```diff
@@ -0,0 +1,739 @@
...
```
Comment (JulianGCalderon) [inline src/libfuncs/bounded_int.rs:310]:
```suggestion
/// Generate MLIR operations for the `bounded_int_mul` libfunc.
```

**Review by JulianGCalderon — APPROVED** [review]
Nice work! Looks good to me. I only found a few documentation errors

**Pull request merged**